lib/cloud/mocks/aws_sts.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mocks
+
+import (
+	"bytes"
+	"html/template"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	awstime "github.com/aws/smithy-go/time"
+	"github.com/gravitational/trace"
+)
+
+// AssumedRoleARN is a sample ARN of an assumed role.
+const AssumedRoleARN = "arn:aws:sts::123456789012:assumed-role/role-name/role-session-name"
+
+// AssumeRoleXMLResponse returns the AssumeRole response in XML format.
+func AssumeRoleXMLResponse(output *sts.AssumeRoleOutput) ([]byte, error) {
+	// AWS SDK uses private functions to serialize/deserialize API structs.
+	// Instead of mimicking the SDK, this output was captured from AWS CLI with
+	// debug flag.
+	tmpl := template.Must(template.New("sts:AssumeRole").Funcs(awsTemplateFuncs).Parse(`
+<AssumeRoleResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
+  <AssumeRoleResult>
+    <AssumedRoleUser>
+      <AssumedRoleId>{{ formatString .AssumedRoleUser.AssumedRoleId }}</AssumedRoleId>
+      <Arn>{{ formatString .AssumedRoleUser.Arn }}</Arn>
+    </AssumedRoleUser>
+    <Credentials>
+      <AccessKeyId>{{ formatString .Credentials.AccessKeyId }}</AccessKeyId>
+      <SecretAccessKey>{{ formatString .Credentials.SecretAccessKey }}</SecretAccessKey>
+      <SessionToken>{{ formatString .Credentials.SessionToken }} </SessionToken>
+      <Expiration>{{ formatTime .Credentials.Expiration }}</Expiration>
+    </Credentials>
+  </AssumeRoleResult>
+  <ResponseMetadata>
+    <RequestId>22222222-2222-2222-2222-222222222222</RequestId>
+  </ResponseMetadata>
+</AssumeRoleResponse>
+`))
```

Comment:
It's quite difficult to use `encoding/xml` with the AWS SDK structs **directly**. For example, the `Expiration` is `*time.time`, to marshal/unmarshal it, the SDK does
https://github.com/aws/aws-sdk-go-v2/blob/2b422c03a22416e47fbf567ee4e909329801476a/service/sts/deserializers.go#L1380-L1395
```
case strings.EqualFold("Expiration", t.Name.Local):
			val, err := decoder.Value()
			if err != nil {
				return err
			}
			if val == nil {
				break
			}
			{
				xtv := string(val)
				t, err := smithytime.ParseDateTime(xtv)
				if err != nil {
					return err
				}
				sv.Expiration = ptr.Time(t)
			}
```
So when I use `encoding/xml` to unmarshal the response in the middleware, the expiration is wrong. We only needed strings and those are fine.

Let me do a proper fix then for both xml marshal/unmarshal.  The easiest option would be using `https://github.com/aws/aws-sdk-go/tree/main/private/protocol/xml/xmlutil`. I will try that first. Otherwise, we have to redefine the SDK structs with custom types. I would rather not do that. Though the `xmlutil` is only in SDK v1.

Reply:
add some **simple** aws XML helper functions. they are not the best but should suffice for now. will need a lot more reflect magic to do it properly.